src/main/java/ch/njol/skript/expressions/ExprEnchantments.java

```diff
@@ -109,36 +124,57 @@ public void change(Event e, @Nullable Object[] delta, ChangeMode mode) {
 		
 		switch (mode) {
 			case ADD:
-				for (ItemType item : source)
-					item.addEnchantments(enchants);
+				if (isStored) {
+					for (ItemType item : source)
+						item.addStoredEnchantments(enchants);
+				} else {
+					for (ItemType item : source)
+						item.addEnchantments(enchants);
+				}
 				break;
 			case REMOVE:
 			case REMOVE_ALL:
 				for (ItemType item : source) {
 					ItemMeta meta = item.getItemMeta();
+					EnchantmentStorageMeta storageMeta = item.getEnchantmentStorageMeta();
 					assert meta != null;
 					for (EnchantmentType enchant : enchants) {
 						Enchantment ench = enchant.getType();
 						assert ench != null;
-						if (enchant.getInternalLevel() == -1
-								|| meta.getEnchantLevel(ench) == enchant.getLevel()) {
-							// Remove directly from meta since it's more efficient on this case
-							meta.removeEnchant(ench);
+						if (enchant.getInternalLevel() == -1)
+							continue;
+						if (isStored && storageMeta != null) {
```

Comment:
same here, i don't feel that having no stored enchantments means we should fallback to removing regular enchantments. that might produce confusing behavior imo